src/Symfony/Component/Serializer/Deserialize/Config/DeserializeConfig.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Deserialize\Config;
+
+/**
+ * Deserialization base configuration.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental in 7.0
+ */
+class DeserializeConfig
+{
+    /**
+     * @var list<string>
+     */
+    protected array $groups = [];
+
+    protected bool $forceGenerateTemplate = false;
+
+    protected ?bool $lazy = null;
+
+    protected string $dateTimeFormat = \DateTimeInterface::RFC3339;
+
+    protected JsonDeserializeConfig $jsonConfig;
+
+    protected CsvDeserializeConfig $csvConfig;
+
+    public function __construct()
+    {
+        $this->jsonConfig = new JsonDeserializeConfig();
+        $this->csvConfig = new CsvDeserializeConfig();
+    }
+
+    /**
+     * @return list<string>
+     */
+    public function groups(): array
+    {
+        return $this->groups;
+    }
+
+    /**
+     * @param non-empty-string|non-empty-array<int, non-empty-string> $groups
+     */
+    public function withGroups(array|string $groups): static
+    {
+        $clone = clone $this;
+        $clone->groups = array_values(array_unique((array) $groups));
```

Comment:
I'm unsure about these configuration objects and would rather see a simple `array $config` (for maintenance and practical reasons). Also, like here, I'd avoid any logic inside metadata objects especially that no matter what this will be called everytime the PHP code is executed.

Reply:
As an user i find it more practical and safer to have an object, it allow to better track change in it, find potential failure with static analysis and a better support into IDE when using this.

However such config into symfony is traditionally represented as an array, as it's simpler when dealing with extensions points (specially if we have 2 libraries that extends this config, we would not be able to pass both of them)

Maybe it could be both views ? : 

 * Use an array in the implementation detail
 * Provide a `DeserializeConfigBuilder` / `DeserializeConfigBuilderInterface` to construct this array (and allow it also in the interface, the implementation will only call the toArray / build method of this interface)

Reply:
Actually, I would have liked to get rid of array for both internal and user land.

Indeed, this allows us to avoid checking during runtime that the context has the required key and that the related value is valid. And it is way better documented, IDE friendly, valid, etc.

I agree that it won't be possible to pass two extensions config at the same time, and this must be indeed fixed. Maybe we can leverage an old context implementation that I tried to introduce in Symfony earlier (but can't because of BC reasons): https://github.com/mtarld/symfony/pull/1/files?

Anyway, I can get that it might be simpler to maintain using arrays. In that case, I'd go with the @joelwurtz point of view. So we might reuse/improve existing [ContextBuilders](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Serializer/Context/ContextBuilderInterface.php)

Reply:
> Maybe we can leverage an old context implementation that I tried to introduce in Symfony earlier (but can't because of BC reasons): https://github.com/mtarld/symfony/pull/1/files?

If we stick to classes, indeed having the base config class final and adding something like : 

For the user : `$config->addExtension('api-platform', $apiPlatformConfig);` 

In the extension implementation : 

```
$apiPlatformSerializerConfig = $config->getExtension('api-platform') ?? new ApiSerializerConfig();

if (!$apiPlatformSerializerConfig instanceof ApiSerializerConfig) {
    throw new \LogicException('api platform extension must be an ...');
}
```

Would work (it's an example, method / name used are just for show), and it may not be forced to be an object (so each library can do what they want at this level, it's their responsibilities)

You could also just make the `SerializerConfig` extends `\ArrayObject` (may be simpler)

Reply:
+1 for simple arrays, especially now that keys and values can be checked and autocompleted by IDEs and static analyzers by defining the array shape in PHPDoc blocks.

For the record @nicolas-grekas, @weaverryan and I started an experiment to transform the whole Symfony config as simple PHP arrays (instead of YAML) during the API Platform Conference.